redhat-appstudio/service-provider-integration-operator
# Compensate the changes in the dependent objects in case of the binding update failure

### What does this PR do?

This deletes the dependent objects (secret(s) and service account(s)) created for a binding if there is a failure during binding reconciliation after those objects have been created.

This makes sure that we don't create "garbage" until all the operations "click" in the right order.

In addition to recovering from binding update failure, this also tries to solve the issue that was especially apparent on OpenShift, which automatically injects an image pull secret to any service account. This immediate update to a newly created SA wasn't accounted for in our code and so this PR adds a retry loop that tries to reload the SA from the cluster and retry the linking (which happens after a create of the SA).

### What issues does this PR fix or reference?
https://issues.redhat.com/browse/SVPI-210
https://issues.redhat.com/browse/SVPI-351

### How to test this PR?
This is especially reproducible on OpenShift with service accounts.

1. Create a binding with a managed service account definition and secret without name so that generated names are used.
2. Provide the data for the token either using OAuth flow or manual upload so that the binding moves to the injected state.
3. observe that there is only 1 new secret and 1 new service account as per definition

```
quay.io/redhat-appstudio/pull-request-builds:spi-operator-445
quay.io/redhat-appstudio/pull-request-builds:spi-oauth-445
```
